Gordon360/Services/HousingService.cs

```diff
@@ -537,5 +560,215 @@ public bool ChangeApplicationDateSubmitted(int applicationID)
             _context.SaveChanges();
             return true;
         }
+
+        /// <summary>
+        /// update roommate imformation
+        /// </summary>
+        /// <param name="username"> The username for the user who complete the aplication form </param>
+        /// <param name="applicantion_id"> The ID of this application </param>
+        /// <param name="emailList"> A list of applicants' emails </param>
+        public async Task UpdateRoommateAsync(string username, string applicantion_id, string[] emailList)
+        {
+            var GordonID = _context.ACCOUNT.FirstOrDefault(a => a.AD_Username == username)?.gordon_id;
+            if (GordonID == null)
+            {
+                throw new ResourceNotFoundException() { ExceptionMessage = "The applicant could not be found" };
+            }
+            var existApplicant = _context.Applicant.Where(x => x.ApplicationID == applicantion_id);
+            var existMaxYear = _context.Year.Where(x => x.ApplicationID == applicantion_id);
+            foreach (Applicant a in existApplicant)
+            {
+                _context.Applicant.Remove(a);
+            }
+            foreach (Year y in existMaxYear)
+            {
+                _context.Year.Remove(y);
+            }
+            List<string> yearList = new List<string>();
+            foreach (string e in emailList)
+            {
+                if (e != "")
+                {
+                    var newApplicant = new Applicant
+                    {
+                        ApplicationID = applicantion_id,
+                        Applicant1 = e
+                    }; ;
+                    await _context.Applicant.AddAsync(newApplicant);
+
+                    var student = _context.Student.FirstOrDefault(x => x.Email == e);
+                    if (student != null)
+                    {
+                        yearList.Add(student.Class);
+                    }
+                }
+            }
+            var newMaxYear = new Year
+            {
+                ApplicationID = applicantion_id,
+                Year1 = yearList.Max()
+            }; ;
+            await _context.Year.AddAsync(newMaxYear);
```

Comment:
We can simplify the logic for getting the max class year of all applicants, and also make it much more performant, by getting all rows from the Student table in one query:
```suggestion
            var maxYear = _context.Student.Where(s => emailList.Contains(s.Email)).Select(s => s.Class).Max();
            await _context.Year.AddAsync(new Year
            {
                ApplicationID = applicantion_id,
                Year1 = maxYear
            });
```

Reply:
Similarly, we can use a bit of LINQ to simplify the Applicant code:
```suggestion
            var newApplicants = emailList
                .Where(e => String.IsNullOrWhiteSpace(e))
                .Select(email => new Applicant
                    {
                        ApplicationID = applicantion_id,
                        Applicant1 = email
                    });
            _context.Applicant.AddRange(newApplicants);
```